task_manager/task_manager.hh

```diff
@@ -0,0 +1,66 @@
+#pragma once
+
+#include "replica/database.hh"
+#include "repair/row_level.hh"
+
+class task_manager {
+public:
+    enum class task_state {
+        RUNNING,
+        BLOCKED,
+        PAUSED,
+        DONE,
+        FAILED
+    };
+    static std::string serialize_state(task_state state) {
+        switch (state) {
+            case task_state::RUNNING: return "RUNNING";
+            case task_state::BLOCKED: return "BLOCKED";
+            case task_state::PAUSED: return "PAUSED";
+            case task_state::DONE: return "DONE";
+            case task_state::FAILED: return "FAILED";
+            default: return "";
+        }
+    }
+    struct task_status {
+        utils::UUID id;
+        std::string type;
+        task_state state;
+        bool is_abortable;
+        gc_clock::time_point start_time;
+        gc_clock::time_point end_time;
+        std::string error;
+        utils::UUID parent_id;
+        uint64_t number;
+        uint64_t shard;
+        std::string keyspace;
+        std::string table;
+        std::string entity;
+        std::string progress_units;
+        float progress_total;
+        float progress_completed;
+    };
+    class task {
+    private:
+        task_status status;
+    public:
+        task_status& get_status() {
+            return status;
+        }
+
+        virtual void abort() = 0;
+    };
+private:
+    std::map<utils::UUID, seastar::shared_ptr<task>> tasks; // TODO Deexie make it map<module, set<task>>
+public:
+    std::map<utils::UUID, seastar::shared_ptr<task>>& get_tasks() {
+        return tasks;
+    }
+    std::vector<std::string> get_modules();
+    // Returns pointer to a task with given id. If task has completed, it is removed from memory. 
+    seastar::shared_ptr<task> retrieve_task(utils::UUID id);
+    // Waits for a task to be completed.
+    seastar::shared_ptr<task> wait_task_no_retrieve(utils::UUID id);
+    // Waits for a task to be completed and removes it from memory.
+    seastar::shared_ptr<task> wait_task(utils::UUID id);
```

Comment:
I don't understand how we can wait here if this is a `sync` method (no `future<>` is returned).

Reply:
You're right, I changed it.